donkeycar/pipeline/sequence.py

```diff
@@ -1,154 +1,90 @@
-
+                 iterable: Union[List[TubRecord],
+                                 'TfmShallowList[XOut, YOut]'],
+                 x_transform: Callable[[Any], XOut],
+                 y_transform: Callable[[Any], YOut]) -> None:
         self.iterable = iterable
         self.x_transform = x_transform
         self.y_transform = y_transform
-        self.iterator = BaseTfmIterator_(
-            iterable=self.iterable, x_transform=self.x_transform, y_transform=self.y_transform)
-
-    def __len__(self):
-        return len(self.iterator)
 
-    def __iter__(self) -> SizedIterator[Tuple[XOut, YOut]]:
-        return BaseTfmIterator_(
-            iterable=self.iterable, x_transform=self.x_transform, y_transform=self.y_transform)
+    def __len__(self) -> int:
+        return len(self.iterable)
 
-    def __next__(self):
-        return next(self.iterator)
+    def __iter__(self) -> _BaseTfmIterator[X, Y, XOut, YOut]:
+        return _BaseTfmIterator(self)
 
+    def transform_item(self, r: Union[TubRecord, Tuple[X, Y]]) \
+            -> Tuple[XOut, YOut]:
+        if isinstance(r, tuple) and len(r) == 2:
+            x, y = r
+            return self.x_transform(x), self.y_transform(y)
+        else:
+            return self.x_transform(r), self.y_transform(r)
 
-class BaseTfmIterator_(Generic[XOut, YOut],  SizedIterator[Tuple[XOut, YOut]]):
-    '''
-    A basic transforming iterator.
-    Do no use this class directly.
-    '''
 
+class Pipeline(Generic[X, Y, XOut, YOut], SizedIterable[Tuple[XOut, YOut]]):
     def __init__(self,
-                 # Losing a bit of type safety here for a common implementation
-                 iterable: Iterable[Any],
-                 x_transform: Callable[[R], XOut],
-                 y_transform: Callable[[R], YOut]) -> None:
-
+                 iterable: Union['Pipeline', List[TubRecord]],
+                 x_transform: Callable[[Any], XOut],
+                 y_transform: Callable[[Any], YOut]) -> None:
         self.iterable = iterable
         self.x_transform = x_transform
         self.y_transform = y_transform
-        self.iterator = iter(self.iterable)
-
-    def __len__(self):
-        return len(self.iterator)
-
-    def __iter__(self) -> SizedIterator[Tuple[XOut, YOut]]:
-        return BaseTfmIterator_(self.iterable, self.x_transform, self.y_transform)
-
-    def __next__(self):
-        record = next(self.iterator)
-        if (isinstance(record, tuple) and len(record) == 2):
-            x, y = record
-            return self.x_transform(x), self.y_transform(y)
-        else:
-            return self.x_transform(record), self.y_transform(record)
-
-
-class TubSequence(Iterable[TubRecord]):
-    def __init__(self, records: List[TubRecord]) -> None:
-        self.records = records
-
-    def __iter__(self) -> SizedIterator[TubRecord]:
-        return TubSeqIterator(self.records)
-
-    def __len__(self):
-        return len(self.records)
-
-    def build_pipeline(self,
```

Comment:
These helper functions are not needed, `TfmShallowList` or `Pipeline` and just be constructed as user objects.